cpp/include/raft/spatial/knn/detail/ivf_flat/bitonic_sort.cuh

```diff
@@ -0,0 +1,231 @@
+
+  /**
+   * You can think of this function in two ways:
+   *
+   *   1) Sort any bitonic sequence.
+   *   2) Merge two halfs of the input data assuming they're already sorted, and their order is
+   *      opposite (i.e. either ascending, descending or vice-versa).
+   *
+   * The input pointers are unique per-thread.
+   * See the class description for the description of the data layout.
+   *
+   * @param keys
+   *   is a device pointer to a contiguous array of keys, unique per thread; must be at least `Size`
+   *   elements long.
+   * @param payloads
+   *   are zero or more associated arrays of the same size as keys, which are sorted together with
+   *   the keys; must be at least `Size` elements long.
+   */
+  template <typename KeyT, typename... PayloadTs>
+  __device__ __forceinline__ void merge(KeyT* __restrict__ keys,
+                                        PayloadTs* __restrict__... payloads) const
+  {
+    return bitonic<Size>::merge_(ascending_, warp_width_, keys, payloads...);
+  }
+
+  /**
+   * Sort the data.
+   * The input pointers are unique per-thread.
+   * See the class description for the description of the data layout.
+   *
+   * @param keys
+   *   is a device pointer to a contiguous array of keys, unique per thread; must be at least `Size`
+   *   elements long.
+   * @param payloads
+   *   are zero or more associated arrays of the same size as keys, which are sorted together with
+   *   the keys; must be at least `Size` elements long.
+   */
+  template <typename KeyT, typename... PayloadTs>
+  __device__ __forceinline__ void sort(KeyT* __restrict__ keys,
+                                       PayloadTs* __restrict__... payloads) const
+  {
+    return bitonic<Size>::sort_(ascending_, warp_width_, keys, payloads...);
+  }
+
+  /**
+   * @brief `merge` variant for the case of one element per thread.
+   *
+   * @param key
+   * @param payload
+   */
+  template <typename KeyT, typename... PayloadTs, int S = Size>
+  __device__ __forceinline__ auto merge(KeyT& __restrict__ key,
+                                        PayloadTs& __restrict__... payload) const
+    -> std::enable_if_t<S == 1, void>  // SFINAE to enable this for Size == 1 only
+  {
+    static_assert(S == Size);
+    return merge(*key, *payload...);
+  }
+
+  /**
+   * @brief `sort` variant for the case of one element per thread.
+   *
+   * @param key
+   * @param payload
+   */
+  template <typename KeyT, typename... PayloadTs, int S = Size>
+  __device__ __forceinline__ auto sort(KeyT& __restrict__ key,
+                                       PayloadTs& __restrict__... payload) const
+    -> std::enable_if_t<S == 1, void>  // SFINAE to enable this for Size == 1 only
+  {
+    static_assert(S == Size);
+    return sort(*key, *payload...);
+  }
+
+ private:
+  const int warp_width_;
+  const bool ascending_;
+
+  template <int AnotherSize>
+  friend class bitonic;
+
+  template <typename KeyT, typename... PayloadTs>
```

Comment:
same here w/ templates. We should strive to be consistent. Though, if we want to forego this discussion / change to get this PR in quickly for 22.04, can you create an issue for it so we don't lose track?

Reply:
Closing this with the reference to the new issue https://github.com/rapidsai/raft/issues/591 .